src/GravitationalClustering.js

```diff
@@ -0,0 +1,186 @@
+        this.alpha = options.alpha || 0.03;
+        this.eps = Number.MAX_VALUE;
+        this.gamma = options.gamma || 0.2;
+        this.dist = options.dist || Distance.euclidean;
+        this.iterations = options.iterations || 100;
+
+        this.particles = [];
+        this.outliers = [];
+    }
+
+    /**
+     * Set the points and corresponding masses for the algorithm
+     * @param {Array} X - Points to create the clusters.
+     * @param {Array} masses - Mass of each point (Optional).
+     */
+    train(X, masses) {
+        if (X.length === 0) {
+            throw new RangeError('The matrix length should be greater than 0.');
+        }
+        if (masses === undefined) {
+            masses = new Array(X.length);
+        }
+
+        this.particles = [];
+        this.outliers = [];
+
+        var dim = X[0].length;
+        var elements = X.length;
+        var minValues = new Array(dim);
+        var maxValues = new Array(dim);
+
+        for (var i = 0; i < dim; ++i) {
+            minValues[i] = Number.MAX_VALUE;
+            maxValues[i] = Number.MIN_VALUE;
+        }
+
+        this.uf = new UnionFind();
+        this.disjointElems = new Array(elements);
+        for (i = 0; i < elements; ++i) {
+            this.disjointElems[i] = this.uf.add(i);
+            var elem = X[i];
+            var mass = masses[i];
+            if (dim !== elem.length) throw new RangeError('The element at position ' + i + 'must have a size of '
+                                                        + dim + ' but has: ' + elem.length);
+
+            for (var j = 0; j < dim; ++j) {
+                var comp = elem[j];
+                minValues[j] = Math.min(minValues[j], comp);
+                maxValues[j] = Math.max(maxValues[j], comp);
+            }
+
+            this.particles.push(new Particle(elem, mass));
+        }
+
+        this.eps = this.gamma * this.dist(maxValues, minValues);
+
+        return this.run(this.iterations);
+    }
+
+    /**
+     * Move two particles closer to each other based on the gravitational constant and the selected distance function.
+     * @param {Particle} particleA
+     * @param {Particle} particleB
+     */
+    move(particleA, particleB) {
+        var deltaA = Particle.moveParticle(particleA, particleB, this.gravitationalConstant, this.dist);
+        for (var i = 0; i < deltaA.length; ++i) {
+            particleA.position[i] += deltaA[i];
+            particleB.position[i] -= deltaA[i];
+        }
+    }
+
+    /**
+     * Merge two particle indices in the X array to the same cluster.
+     * @param {DisjointSetNode} a
+     * @param {DisjointSetNode} b
+     */
+    merge(a, b) {
+        this.uf.union(a, b);
+    }
+
+    /**
+     * Run the algorithm n-iterations partially and return the obtained clusters and outliers.
+     * @param {number} iterations - Partial iterations.
+     * @return {object}
+     */
+    run(iterations) {
```

Comment:
yes, I will allow that method, is useful if you disagree with the current model, I'm thinking about predict because is not specified in the algorithm